quic/api/QuicTransportBase.cpp

```diff
@@ -1638,7 +1638,8 @@ QuicTransportBase::notifyPendingWriteOnStream(StreamId id, WriteCallback* wcb) {
   }
   // Add the callback to the pending write callbacks so that if we are closed
   // while we are scheduled in the loop, the close will error out the callbacks.
-  if (!pendingWriteCallbacks_.emplace(id, wcb).second) {
+  auto existing_wcb_pair = pendingWriteCallbacks_.emplace(id, wcb);
+  if (!existing_wcb_pair.second && (existing_wcb_pair.first)->second != wcb) {
     return folly::makeUnexpected(LocalErrorCode::INVALID_WRITE_CALLBACK);
   }
   runOnEvbAsync([id](auto self) {
```

Comment:
Although safe, this will place an unnecessary lambda on evb here. So we would like to understand why you want to do it first.

Reply:
Passing `nullptr` on subsequent calls will still cause INVALID_WRITE_CALLBACK to be thrown, correct? The same if condition with the emplace() call will be triggered.

It depends on what the semantics of the call should be: The comment says: 

```
Inform the transport that there is data to write on a given stream.
```

So it's unclear if this is valid for one write on the stream (i.e., one call to `onStreamWriteReady` will cause this callback to be cleared), or if the callback is only set once per stream.

If the callback is only set once per stream, but subsequent writes require notifyPendingWriteOnStream to be called to invoke onStreamWriteReady, how would these subsequent writes work? They just silently error, and we expect the error to be discarded?

If it's not necessary to call notifyPendingWriteOnStream for each write, then perhaps the name of this function should be changed to "registerWriteCallback" or something, but I'm not sure how the runtime would know to call onStreamWriteReady.

Hopefully that explains why I am a bit confused :)

Reply:
we uninstall the write callback before invoke the callback:
https://github.com/facebookincubator/mvfst/blob/master/quic/api/QuicTransportBase.cpp#L1666
https://github.com/facebookincubator/mvfst/blob/master/quic/api/QuicTransportBase.cpp#L1412
https://github.com/facebookincubator/mvfst/blob/master/quic/api/QuicTransportBase.cpp#L1467

Reply:
So when you call notifyPendingWriteOnStream the first time for a stream id X, we will schedule a lambda on the evb which will check if you can still write more, and if so, we will uninstall your callback and give you a onStreamWriteReady. Then you can notifyPendingWriteonStream again with the same ID and same callback and there won't be no error. What you cannot do, is to notifyPendingWriteOnStream with the same ID and same callback again and again within the same event loop. I guess I can see an argument saying inn your onStreamWriteReady code you may need to collect multiple sources of data and each of them needs a notifyPendingWriteOnStream with the same stream ID and callback. But I guess you can have your layer gathering them first then just notifyPendingWriteOnStream once?

Let me know if this makes sense.

Reply:
Yes, that makes sense.

In this case, maybe it would be good to have a different error message instead of `INVALID_WRITE_CALLBACK`, e.g., `ALREADY_INSTALLED` - then, depending on the semantics of the calling code, a returned `ALREADY_INSTALLED` error can be safely ignored and there are no extra lambdas.
This way, the callback-installed state doesn't have to be duplicated between mvfst and the caller.

If something like this would be ok, I'm happy to revise the PR

Reply:
I guess that makes sense. But then the question then becomes what's the right error code for the connection write callback case? We return INVALID_WRITE_CALLBACK for connection error callback as well when user tries to install a write callback to a connection that already has one. Maybe that should also be a ALREADY_INSTALLED?

Reply:
Let me know if the new version above answers your question?